src/pumps/MultiFlowPump.sol

```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
      */
-    function readCappedReserves(address well) external view returns (uint256[] memory cappedReserves) {
+    function readCappedReserves(
+        address well,
+        bytes calldata data
+    ) external view returns (uint256[] memory cappedReserves) {
+        (, uint256 capInterval, CapReservesParameters memory crp) =
+            abi.decode(data, (bytes16, uint256, CapReservesParameters));
         bytes32 slot = _getSlotForAddress(well);
         uint256[] memory currentReserves = IWell(well).getReserves();
-        (uint8 numberOfReserves, uint40 lastTimestamp, bytes16[] memory lastReserves) = slot.readLastReserves();
+        uint8 numberOfReserves;
+        uint40 lastTimestamp;
+        (numberOfReserves, lastTimestamp, cappedReserves) = slot.readLastReserves();
         if (numberOfReserves == 0) {
             revert NotInitialized();
         }
         uint256 deltaTimestamp = _getDeltaTimestamp(lastTimestamp);
-        cappedReserves = new uint256[](numberOfReserves);
         if (deltaTimestamp == 0) {
-            for (uint256 i; i < numberOfReserves; ++i) {
-                cappedReserves[i] = lastReserves[i].pow_2ToUInt();
-            }
             return cappedReserves;
         }
 
-        bytes16 capExponent = ((deltaTimestamp - 1) / CAP_INTERVAL + 1).fromUInt();
+        uint256 capExponent = ((deltaTimestamp - 1) / capInterval + 1);
+        cappedReserves = _capReserves(well, cappedReserves, currentReserves, capExponent, crp);
+    }
+
+    function _capReserves(
+        address well,
+        uint256[] memory lastReserves,
+        uint256[] memory reserves,
+        uint256 capExponent,
+        CapReservesParameters memory crp
+    ) internal view returns (uint256[] memory cappedReserves) {
+        // Assume two token well
+        if (reserves.length != 2) {
+            revert TooManyTokens();
+        }
+
+        cappedReserves = new uint256[](2);
+        cappedReserves[0] = reserves[0];
+        cappedReserves[1] = reserves[1];
 
-        for (uint256 i; i < numberOfReserves; ++i) {
-            cappedReserves[i] =
-                _capReserve(lastReserves[i], currentReserves[i].fromUIntToLog2(), capExponent).pow_2ToUInt();
+        Call memory wf = IWell(well).wellFunction();
+        IMultiFlowPumpWellFunction mfpWf = IMultiFlowPumpWellFunction(wf.target);
+
+        cappedReserves = _capLpTokenSupply(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data, true);
+
+        if (cappedReserves.length == 0) {
+            cappedReserves = _capRatios(lastReserves, reserves, capExponent, crp, mfpWf, wf.data);
+
+            cappedReserves = _capLpTokenSupply(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data, false);
+        } else {
+            cappedReserves = _capRatios(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data);
         }
     }
 
+    struct CapRatiosVariables {
```

Comment:
can we move this up

Reply:
Fixed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/4a2f50ee54b42155085ec2f01baadb63bc41eccc